README.md
# Stark Bank .NET SDK

Welcome to the Stark Bank .NET SDK! This tool is made for .NET
developers who want to easily integrate with our API.
This SDK version is compatible with the Stark Bank API v2.

If you have no idea what Stark Bank is, check out our [website](https://www.starkbank.com/)
and discover a world where receiving or making payments
is as easy as sending a text message to your client!

## Supported .NET Versions

This library supports the following .NET versions:

* .NET Standard 2.0+

## Stark Bank API documentation

Feel free to take a look at our [API docs](https://www.starkbank.com/docs/api).

## Versioning

This project adheres to the following versioning pattern:

Given a version number MAJOR.MINOR.PATCH, increment:

- MAJOR version when the **API** version is incremented. This may include backwards incompatible changes;
- MINOR version when **breaking changes** are introduced OR **new functionalities** are added in a backwards compatible manner;
- PATCH version when backwards compatible bug **fixes** are implemented.

## Setup

### 1. Install our SDK

StarkBank`s .NET SDK is available on NuGet as starkbank 2.2.3.

1.1 To install the Package Manager:

```sh
Install-Package starkbank -Version 2.2.3
```

1.2 To install the .NET CLI:

```sh
dotnet add package starkbank --version 2.2.3
```

1.3 To install by PackageReference:

```sh
<PackageReference Include="starkbank" Version="2.2.3" />
```

1.4 To install with Paket CLI:

```sh
paket add starkbank --version 2.2.3
```

### 2. Create your Private and Public Keys

We use ECDSA. That means you need to generate a secp256k1 private
key to sign your requests to our API, and register your public key
with us so we can validate those requests.

You can use one of following methods:

2.1. Check out the options in our [tutorial](https://starkbank.com/faq/how-to-create-ecdsa-keys).

2.2. Use our SDK:

```c#
(string privateKey, string publicKey) = StarkBank.Key.Create();

# or, to also save .pem files in a specific path
(string privateKey, string publicKey) = StarkBank.Key.Create("file/keys");
```

**NOTE**: When you are creating a new integration user, it is recommended that you create the
keys inside the infrastructure that will use it, in order to avoid risky internet
transmissions of your **private-key**. Then you can export the **public-key** alone to the
computer where it will be used in the new Project creation.

### 3. Register your user credentials

You can interact directly with our API using two types of users: Projects and Organizations.

- **Projects** are workspace-specific users, that is, they are bound to the workspaces they are created in.
One workspace can have multiple Projects.
- **Organizations** are general users that control your entire organization.
They can control all your Workspaces and even create new ones. The Organization is bound to your company's tax-ID only.
Since this user is unique in your entire organization, only one credential can be linked to it.


3.1 To create a Project in Sandbox:

3.1.1. Log into [Starkbank Sandbox](https://sandbox.web.starkbank.com)

3.1.2. Go to Menu > Projects

3.1.3. Create a Project: Give it a name and upload the public key you created in section 2.

3.1.4. After creating the Project, get its Project ID

3.1.5. Use the Project ID and private key to create the object below:

```c#
// Get your private key from an environment variable or an encrypted database.
// This is only an example of a private key content. You should use your own key.
string privateKeyContent = "-----BEGIN EC PARAMETERS-----\nBgUrgQQACg==\n-----END EC PARAMETERS-----\n-----BEGIN EC PRIVATE KEY-----\nMHQCAQEEIMCwW74H6egQkTiz87WDvLNm7fK/cA+ctA2vg/bbHx3woAcGBSuBBAAK\noUQDQgAE0iaeEHEgr3oTbCfh8U2L+r7zoaeOX964xaAnND5jATGpD/tHec6Oe9U1\nIF16ZoTVt1FzZ8WkYQ3XomRD4HS13A==\n-----END EC PRIVATE KEY-----";

StarkBank.Project project = new StarkBank.Project(
    environment: "sandbox",
    id: "5656565656565656",
    privateKey: privateKeyContent
);
```

3.2 While this feature is in beta, to register your Organization's public key, a legal representative of your organization must send an e-mail with the desired public key to developers@starkbank.com. Don`t worry, this flow will soon be integrated with our website. Here is an example on how to handle your Organization in the SDK:

```c#
// Get your private key from an environment variable or an encrypted database.
// This is only an example of a private key content. You should use your own key.
string privateKeyContent = "-----BEGIN EC PARAMETERS-----\nBgUrgQQACg==\n-----END EC PARAMETERS-----\n-----BEGIN EC PRIVATE KEY-----\nMHQCAQEEIMCwW74H6egQkTiz87WDvLNm7fK/cA+ctA2vg/bbHx3woAcGBSuBBAAK\noUQDQgAE0iaeEHEgr3oTbCfh8U2L+r7zoaeOX964xaAnND5jATGpD/tHec6Oe9U1\nIF16ZoTVt1FzZ8WkYQ3XomRD4HS13A==\n-----END EC PRIVATE KEY-----";

StarkBank.Organization organization = new StarkBank.Organization(
    environment: "sandbox",
    id: "5656565656565656",
    privateKey: privateKeyContent,
    workspaceID: null  // You only need to set the workspaceID when you are operating a specific workspaceID
);

// To dynamically use your organization credentials in a specific workspaceID,
// you can use the Organization.WithWorkspace() method:
StarkBank.Balance.Get(user: organization.WithWorkspace("4848484848484848"));
```

NOTE 1: Never hard-code your private key. Get it from an environment variable or an encrypted database.

NOTE 2: We support `'sandbox'` and `'production'` as environments.

NOTE 3: The credentials you registered in `sandbox` do not exist in `production` and vice versa.


### 4. Setting up the user

There are two kinds of users that can access our API: **Project** and **Member**.

- `Member` is the one you use when you log into our webpage with your e-mail.
- `Project` is designed for integrations and is the one meant for our SDK.

There are two ways to inform the user to the SDK:

4.1 Passing the user as argument in all functions:

```c#
StarkBank.Balance balance = StarkBank.Balance.Get(user: project);
```

4.2 Set it as a default user in the SDK:

```c#
StarkBank.Settings.User = project;

StarkBank.Balance balance = StarkBank.Balance.Get();
```

Just select the way of passing the project user that is more convenient to you.
On all following examples we will assume a default user has been set.

### 5. Setting up the error language

The error language can also be set in the same way as the default user:

```c#
StarkBank.Settings.Language = "en-US";
```

Language options are "en-US" for english and "pt-BR" for brazilian portuguese. English is default.

## Testing in Sandbox

Your initial balance is zero. For many operations in Stark Bank, you"ll need funds
in your account, which can be added to your balance by creating a Boleto.

In the Sandbox environment, 90% of the created Boletos will be automatically paid,
so there"s nothing else you need to do to add funds to your account. Just create
a few and wait around a bit.

In Production, you (or one of your clients) will need to actually pay this Boleto
for the value to be credited to your account.

## Usage

Here are a few examples on how to use the SDK. If you have any doubts, check out
the function or class docstring to get more info or go straight to our [API docs].

### Get balance

To know how much money you have in your workspace, run:

```c#
using System;

StarkBank.Balance balance = StarkBank.Balance.Get();

Console.WriteLine(balance);
```

### Get dict key

You can get PIX key's parameters by its id.

```c#
using System;

StarkBank.DictKey dictKey = DictKey.Get("tony@starkbank.com");

Console.WriteLine(dictKey);
```

### Query your DICT keys

To take a look at the PIX keys linked to your workspace, just run the following:

```c#
using System;
using System.Collections.Generic;

IEnumerable<StarkBank.DictKey> dictKeys = StarkBank.DictKey.Query(
    status: "registered",
    type: "evp"
);

foreach(StarkBank.DictKey dictKey in dictKeys) {
    Console.WriteLine(dictKey);
}
```

### Create invoices

You can create dynamic QR Code invoices to charge customers or to receive money from accounts
you have in other banks.

```c#
# coding: utf-8
using System;
using System.Collections.Generic;

List<StarkBank.Invoice> invoices = StarkBank.Invoice.Create(
    new List<StarkBank.Invoice> {
        new StarkBank.Invoice(
            amount: 248,
            descriptions: new List<Dictionary<string, object>>() {
                new Dictionary<string, object> {
                    {"key", "Arya"},
                    {"value", "Not today"}
                }
            },
            discounts: new List<Dictionary<string, object>>() {
                new Dictionary<string, object> {
                    {"percentage", 10},
                    {"due", new DateTime(2021, 4, 25, 13, 17, 15)}
                }
            },
            due: new DateTime(2021, 5, 25, 13, 17, 15),
            expiration: 123456789,
            fine: 2.5,
            interest: 1.3,
            name: "Arya Stark",
            tags: new List<string> { "New sword", "Invoice #1234" },
            taxID: "29.176.331/0001-69"
        )
    }
);

foreach(StarkBank.Invoice invoice in invoices) {
    Console.WriteLine(invoice);
}
```

**Note**: Instead of using Invoice objects, you can also pass each invoice element in dictionary format

### Get an invoice

After its creation, information on an invoice may be retrieved by its id.
Its status indicates whether it's been paid.

```c#
using System;

StarkBank.Invoice invoice = StarkBank.Invoice.Get("5715709195714560");

Console.WriteLine(invoice);
```

### Get an invoice QRCode

After its creation, an invoice QRCode may be retrieved by its id.

```c#
using System;

byte[] png = StarkBank.Invoice.Qrcode("5715709195714560");

System.IO.File.WriteAllBytes("qrcode.png", png);
```

Be careful not to accidentally enforce any encoding on the raw pdf content,
as it may yield abnormal results in the final file, such as missing images
and strange characters.

### Get an invoice PDF

After its creation, an invoice PDF may be retrieved by its id.

```c#
using System;

byte[] pdf = StarkBank.Invoice.Pdf("5715709195714560");

System.IO.File.WriteAllBytes("invoice.pdf", pdf);
```

Be careful not to accidentally enforce any encoding on the raw pdf content,
as it may yield abnormal results in the final file, such as missing images
and strange characters.

### Cancel an invoice

You can also cancel an invoice by its id.
Note that this is not possible if it has been paid already.

```c#
using System;

StarkBank.Invoice invoice = StarkBank.Invoice.Update(
    "6312789471657984",
    status: "canceled"
);

Console.WriteLine(invoice);
```

### Update an invoice

You can update an invoice's amount, due date and expiration by its id.
Note that this is not possible if it has been paid already.

```c#
using System;

StarkBank.Invoice invoice = StarkBank.Invoice.Update(
    "6312789471657984",
    amount: 99999
);

Console.WriteLine(invoice);
```

### Query invoices

You can get a list of created invoices given some filters.

```c#
using System;

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to the following versioning pattern:

Given a version number MAJOR.MINOR.PATCH, increment:

- MAJOR version when the **API** version is incremented. This may include backwards incompatible changes;
- MINOR version when **breaking changes** are introduced OR **new functionalities** are added in a backwards compatible manner;
- PATCH version when backwards compatible bug **fixes** are implemented.


## [Unreleased]
### Added
- Organization user
- Workspace resource

## [2.2.3] - 2020-11-18
### Fixes
- Fixed bad midnight DateTime string conversions

## [2.2.2] - 2020-11-17
### Fixed
- DateTime string timezone bug in some environments

## [2.2.1] - 2020-11-17
### Fixed
- Invoice optional due parameter

## [2.2.0] - 2020-11-16
### Added
- Invoice resource to load your account with dynamic QR Codes
- DictKey resource to get DICT (PIX) key parameters
- Deposit resource to receive transfers passively
- PIX support in Transfer resource
- BrcodePayment support to pay static and dynamic PIX QR Codes

## [2.1.0] - 2020-10-28
### Added
- BoletoHolmes to investigate boleto status according to CIP

## [2.0.0] - 2020-10-20
### Added
- ids parameter to Transaction.query
- ids parameter to Transfer.query
- PaymentRequest resource to pass payments through manual approval flow

### Fixed
- BoletoPayment test case
- UtilityPayment test case

## [0.9.0] - 2020-08-12
### Added
- Transfer.scheduled parameter to allow Transfer scheduling
- Transfer.Delete to cancel scheduled Transfers
- Transaction query by tags

## [0.8.0] - 2020-06-05
### Added
- Transfer query taxID parameter

## [0.7.0] - 2020-05-26
### Added
- StarkBank.Settings.Language to specify error language as "en-US" or "pt-BR"
- Boleto PDF layout option
### Change
- Test user credentials to environment variable instead of hard-code
- Default user from StarkBank.User.Default to StarkBank.Settings.User

## [0.6.0] - 2020-05-12
### Added
- "receiver_name" & "receiver_tax_id" properties to Boleto entities

## [0.5.1] - 2020-05-12
### Fixed
- /GET requests in .NET Framework 4.6.1

## [0.5.0] - 2020-05-12
### Added
- Support for .NET Standard 2.0

## [0.4.0] - 2020-05-05
### Added
- Support for dictionaries in create methods
- "balance" property to Transaction entities
### Fixed
- Docstrings

## [0.3.1] - 2020-05-05
### Fixed
- SDK User-Agent header

## [0.3.0] - 2020-04-29
### Added
- "discounts" property to Boleto entities
### Changed
- Internal folder structure
### Fixed
- pt-BR environment credential headers bug

## [0.2.0] - 2020-04-17
### Changed
- StarkBank.Error.Error to StarkBank.Error.ErrorElement

## [0.1.0] - 2020-04-17
### Added
- Full Stark Bank API v2 compatibility


StarkBank/StarkBank/Invoice/Invoice.cs
using System;
using System.Linq;
using System.Collections.Generic;


namespace StarkBank
{
    /// <summary>
    /// Invoice object
    /// <br/>
    /// When you initialize an Invoice, the entity will not be automatically
    /// sent to the Stark Bank API. The 'create' function sends the objects
    /// to the Stark Bank API and returns the list of created objects.
    /// <br/>
    /// Properties:
    /// <list>
    ///     <item>ID [string, default null]: unique id returned when Invoice is created. ex: "5656565656565656"</item>
    ///     <item>Name [string]: payer name. ex: "Iron Bank S.A."</item>
    ///     <item>TaxID [string]: payer tax ID (CPF or CNPJ) with or without formatting. ex: "01234567890" or "20.018.183/0001-80"</item>
    ///     <item>Due [DateTime, default now + 2 days]: Invoice due date in UTC ISO format. ex: DateTime(2020, 3, 10, 10, 30, 12, 15)</item>
    ///     <item>Expiration [long integer, default null]: time interval in seconds between due date and expiration date. ex 123456789</item>
    ///     <item>Fine [float, default 2.0]: Invoice fine for overdue payment in %. ex: 2.5</item>
    ///     <item>Interest [float, default 1.0]: Invoice monthly interest for overdue payment in %. ex: 5.2</item>
    ///     <item>Descriptions [list of dictionaries, default null]: list of dictionaries with "key":string and (optional) "value":string pairs. ex: new List<Dictionary<string,string>>(){new Dictionary<string, string>{{"key", "Taxes"},{"value", "100"}}</item>
    ///     <item>Discounts [list of dictionaries, default null]: list of dictionaries with "percentage":float and "due":string pairs. ex: new List<Dictionary<string,string>>(){new Dictionary<string, string>{{"percentage", 1.5},{"due", "2020-11-25T17:59:26.249976+00:00"}}</item>
    ///     <item>Tags [list of strings, default null]: list of strings for tagging</item>
    ///     <item>NominalAmount [long integer, default null]: Invoice emission value in cents (will change if invoice is updated, but not if it's paid). ex: 400000</item>
    ///     <item>Amount [long integer]: Invoice value in cents. Minimum = 0 (any value will be accepted). ex: 1234 (= R$ 12.34)</item>
    ///     <item>FineAmount [long integer, default null]: Invoice fine value calculated over NominalAmount. ex: 20000</item>
    ///     <item>InterestAmount [long integer, default null]: Invoice interest value calculated over NominalAmount. ex: 10000</item>
    ///     <item>DiscountAmount [long integer, default null]: Invoice discount value calculated over NominalAmount. ex: 3000</item>
    ///     <item>Brcode [string, default null]: BR Code for the Invoice payment. ex: "00020101021226800014br.gov.bcb.pix2558invoice.starkbank.com/f5333103-3279-4db2-8389-5efe335ba93d5204000053039865802BR5913Arya Stark6009Sao Paulo6220051656565656565656566304A9A0"</item>
    ///     <item>Fee [integer, default null]: fee charged by this Invoice. ex: 65 (= R$ 0.65)</item>
    ///     <item>Status [string, default null]: current Invoice status. ex: "created", "paid", "canceled" or "overdue"</item>
    ///     <item>Created [DateTime, default null]: creation datetime for the Invoice. ex: DateTime(2020, 3, 10, 10, 30, 0, 0)</item>
    ///     <item>Updated [DateTime, default null]: latest update datetime for the Invoice. ex: DateTime(2020, 3, 10, 10, 30, 0, 0)</item>
    /// </list>
    /// </summary>
    public partial class Invoice : Utils.Resource
    {
        public string Name { get; }
        public string TaxID { get; }
        public DateTime? Due { get; }
        public long? Expiration { get; }
        public double? Fine { get; }
        public double? Interest { get; }
        public List<Dictionary<string, object>> Descriptions { get; }
        public List<Dictionary<string, object>> Discounts { get; }
        public List<string> Tags { get; }
        public long Amount { get; }
        public long? NominalAmount { get; }
        public long? FineAmount { get; }
        public long? InterestAmount { get; }
        public long? DiscountAmount { get; }
        public string Brcode { get; }
        public int? Fee { get; }
        public string Status { get; }
        public DateTime? Created { get; }
        public DateTime? Updated { get; }

        /// <summary>
        /// Invoice object
        /// <br/>
        /// When you initialize an Invoice, the entity will not be automatically
        /// sent to the Stark Bank API. The 'create' function sends the objects
        /// to the Stark Bank API and returns the list of created objects.
        /// <br/>
        /// Parameters (required):
        /// <list>
        ///     <item>amount [long integer]: Invoice value in cents. Minimum = 0 (any value will be accepted). ex: 1234 (= R$ 12.34)</item>
        ///     <item>name [string]: payer name. ex: "Iron Bank S.A."</item>
        ///     <item>taxID [string]: payer tax ID (CPF or CNPJ) with or without formatting. ex: "01234567890" or "20.018.183/0001-80"</item>
        /// </list>
        /// <br/>
        /// Parameters (optional):
        /// <list>
        ///     <item>due [DateTime, default now + 2 days]: Invoice due date in UTC ISO format. ex: DateTime(2020, 3, 10, 10, 30, 0, 0)</item>
        ///     <item>expiration [long integer, default null]: time interval in seconds between due date and expiration date. ex 123456789</item>
        ///     <item>fine [float, default 2.0]: Invoice fine for overdue payment in %. ex: 2.5</item>
        ///     <item>interest [float, default 1.0]: Invoice monthly interest for overdue payment in %. ex: 5.2</item>
        ///     <item>descriptions [list of dictionaries, default null]: list of dictionaries with "key":string and (optional) "value":string pairs. ex: new List<Dictionary<string,string>>(){new Dictionary<string, string>{{"key", "Taxes"},{"value", "100"}}</item>
        ///     <item>discounts [list of dictionaries, default null]: list of dictionaries with "percentage":float and "due":string pairs. ex: new List<Dictionary<string,object>>(){new Dictionary<string, string>{{"percentage", 1.5},{"due", DateTime(2020, 3, 10, 10, 30, 12, 15)}}</item>
        ///     <item>tags [list of strings, default null]: list of strings for tagging</item>
        /// </list>
        /// <br/>
        /// Attributes (return-only):
        /// <list>
        ///     <item>nominalAmount [long integer, default null]: Invoice emission value in cents (will change if invoice is updated, but not if it's paid). ex: 400000</item>
        ///     <item>fineAmount [long integer, default null]: Invoice fine value calculated over nominalAmount. ex: 20000</item>
        ///     <item>interestAmount [long integer, default null]: Invoice interest value calculated over nominalAmount. ex: 10000</item>
        ///     <item>discountAmount [long integer, default null]: Invoice discount value calculated over nominalAmount. ex: 3000</item>
        ///     <item>id [string, default null]: unique id returned when Invoice is created. ex: "5656565656565656"</item>
        ///     <item>brcode [string, default null]: BR Code for the Invoice payment. ex: "00020101021226800014br.gov.bcb.pix2558invoice.starkbank.com/f5333103-3279-4db2-8389-5efe335ba93d5204000053039865802BR5913Arya Stark6009Sao Paulo6220051656565656565656566304A9A0"</item>
        ///     <item>fee [integer, default null]: fee charged by this Invoice. ex: 65 (= R$ 0.65)</item>
        ///     <item>status [string, default null]: current Invoice status. ex: "created", "paid", "canceled" or "overdue"</item>
        ///     <item>created [DateTime, default null]: creation datetime for the Invoice. ex: DateTime(2020, 3, 10, 10, 30, 0, 0)</item>
        ///     <item>updated [DateTime, default null]: latest update datetime for the Invoice. ex: DateTime(2020, 3, 10, 10, 30, 0, 0)</item>
        /// </list>
        /// </summary>
        public Invoice(long amount, string name, string taxID, DateTime? due = null, long? expiration = null, double? fine = null, double? interest = null,
            List<string> tags = null, List<Dictionary<string, object>> descriptions = null, List<Dictionary<string, object>> discounts = null,
            long? nominalAmount = null, long? fineAmount = null, long? interestAmount = null, long? discountAmount = null,
            string id = null, string brcode = null, int? fee = null, string status = null, DateTime? created = null, DateTime? updated = null) : base(id)
        {
            Amount = amount;
            Name = name;
            TaxID = taxID;
            Expiration = expiration;
            Due = due;
            Fine = fine;
            Interest = interest;
            Tags = tags;
            Descriptions = descriptions;
            Discounts = discounts;
            NominalAmount = nominalAmount;
            FineAmount = fineAmount;
            InterestAmount = interestAmount;
            DiscountAmount = discountAmount;
            Brcode = brcode;
            Fee = fee;
            Status = status;
            Created = created;
            Updated = updated;
        }

        internal new Dictionary<string, object> ToJson()
        {
            Dictionary<string, object> json = base.ToJson();
            json["Due"] = new Utils.StarkBankDateTime((DateTime) json["Due"]);
            return json;
        }

        /// <summary>
        /// Create Invoices
        /// <br/>
        /// Send a list of Invoice objects for creation in the Stark Bank API
        /// <br/>
        /// Parameters (required):
        /// <list>
        ///     <item>Invoices [list of Invoice objects]: list of Invoice objects to be created in the API</item>
        /// <br/>
        /// </list>
        /// Parameters (optional):
        /// <list>
        ///     <item>user [Organization/Project object]: Organization or Project object. Not necessary if StarkBank.Settings.User was set before function call</item>
        /// </list>
        /// <br/>
        /// Return:
        /// <list>
        ///     <item>list of Invoice objects with updated attributes</item>
        /// </list>
        /// </summary>
        public static List<Invoice> Create(List<Invoice> invoices, User user = null)
        {
            (string resourceName, Utils.Api.ResourceMaker resourceMaker) = Resource();
            return Utils.Rest.Post(
                resourceName: resourceName,
                resourceMaker: resourceMaker,
                entities: invoices,
                user: user
            ).ToList().ConvertAll(o => (Invoice)o);
        }

        /// <summary>
        /// Create Invoices
        /// <br/>
        /// Send a list of Invoice objects for creation in the Stark Bank API
        /// <br/>
        /// Parameters (required):
        /// <list>
        ///     <item>invoices [list of Dictionaries]: list of Dictionaries representing the Invoices to be created in the API</item>
        /// <br/>
        /// </list>
        /// Parameters (optional):
        /// <list>
        ///     <item>user [Organization/Project object]: Organization or Project object. Not necessary if StarkBank.Settings.User was set before function call</item>
        /// </list>
        /// <br/>
        /// Return:
        /// <list>
        ///     <item>list of Invoice objects with updated attributes</item>
        /// </list>
        /// </summary>
        public static List<Invoice> Create(List<Dictionary<string, object>> invoices, User user = null)
        {
            (string resourceName, Utils.Api.ResourceMaker resourceMaker) = Resource();
            return Utils.Rest.Post(
                resourceName: resourceName,
                resourceMaker: resourceMaker,
                entities: invoices,
                user: user
            ).ToList().ConvertAll(o => (Invoice)o);
        }

        /// <summary>
        /// Retrieve a specific Invoice
        /// <br/>
        /// Receive a single Invoice object previously created in the Stark Bank API by passing its id
        /// <br/>
        /// Parameters(required):
        /// <list>
        ///     <item>id[string]: object unique id. ex: "5656565656565656"</item>
        /// </list>
        /// <br/>
        /// Parameters(optional):
        /// <list>
        ///     <item>user[Project object]: Project object. Not necessary if StarkBank.User.Default was set before function call</item>
        /// </list>
        /// <br/>
        /// Return:
        /// <list>
        ///     <item>Invoice object with updated attributes</item>
        /// </list>
        /// </summary>
        public static Invoice Get(string id, User user = null)
        {
            (string resourceName, Utils.Api.ResourceMaker resourceMaker) = Resource();
            return Utils.Rest.GetId(
                resourceName: resourceName,
                resourceMaker: resourceMaker,
                id: id,
                user: user
            ) as Invoice;
        }

        /// <summary>
        /// Retrieve a specific Invoice pdf file
        /// <br/>
        /// Receive a single Invoice pdf receipt file generated in the Stark Bank API by passing its id.
        /// <br/>
        /// Parameters (required):
        /// <list>
        ///     <item>id [string]: object unique id. ex: "5656565656565656"</item>
        /// </list>
        /// <br/>
        /// Parameters (optional):
        /// <list>
        ///     <item>user [Organization/Project object]: Organization or Project object. Not necessary if StarkBank.Settings.User was set before function call</item>
        /// </list>
        /// <br/>
        /// Return:
        /// <list>
        ///     <item>Invoice pdf file</item>
        /// </list>
        /// </summary>
        public static byte[] Pdf(string id, User user = null)
        {
        (string resourceName, Utils.Api.ResourceMaker resourceMaker) = Resource();
        return Utils.Rest.GetPdf(
            resourceName: resourceName,
            resourceMaker: resourceMaker,
            id: id,
            user: user
        );
        }

        /// <summary>
        /// Retrieve Invoices
        /// <br/>
        /// Receive an IEnumerable of Invoice objects previously created in the Stark Bank API
        /// <br/>
        /// Parameters (optional):
        /// <list>
        ///     <item>limit [integer, default null]: maximum number of objects to be retrieved. Unlimited if null. ex: 35</item>
        ///     <item>after [DateTime, default null] date filter for objects created only after specified date. ex: DateTime(2020, 3, 10)</item>
        ///     <item>before [DateTime, default null] date filter for objects created only before specified date. ex: DateTime(2020, 3, 10)</item>
        ///     <item>status [string, default null]: filter for status of retrieved objects. ex: "created", "paid", "canceled" or "overdue"</item>
        ///     <item>tags [list of strings, default null]: tags to filter retrieved objects. ex: ["tony", "stark"]</item>
        ///     <item>ids [list of strings, default null]: list of ids to filter retrieved objects. ex: ["5656565656565656", "4545454545454545"]</item>
        ///     <item>user [Project object, default null]: Project object. Not necessary if StarkBank.User.Default was set before function call</item>
        /// </list>
        /// <br/>
        /// Return:
        /// <list>
        ///     <item>IEnumerable of Invoice objects with updated attributes</item>
        /// </list>
        /// </summary>
        public static IEnumerable<Invoice> Query(int? limit = null, DateTime? after = null, DateTime? before = null,
            string status = null, List<string> tags = null, List<string> ids = null, User user = null)
        {
            (string resourceName, Utils.Api.ResourceMaker resourceMaker) = Resource();
            return Utils.Rest.GetList(
                resourceName: resourceName,
                resourceMaker: resourceMaker,
                query: new Dictionary<string, object> {
                    { "limit", limit },
                    { "after", new Utils.StarkBankDate(after) },
                    { "before", new Utils.StarkBankDate(before) },
                    { "status", status },
                    { "tags", tags },
                    { "ids", ids }
                },
                user: user
            ).Cast<Invoice>();
        }

        /// <summary>
        /// Update Invoice entity
        /// <br/>
        /// Update an Invoice by passing id, if it hasn't been paid yet.
        /// <br/>
        /// Parameters(required):
        /// <list>
        ///     <item>id[string]: object unique id. ex: "5656565656565656"</item>
        /// </list>
        /// <br/>
        /// Parameters (optional):
        /// <list>
        ///     <item>status [string]: you may cancel the Invoice by passing "canceled" in the status. ex: "canceled"</item>
        ///     <item>amount [long integer]: nominal amount charged by the Invoice. ex: 100 (R$1.00)</item>
        ///     <item>due [DateTime, default today + 2 days]: Invoice due date in UTC ISO format. ex: DateTime(2020, 3, 10, 10, 30, 12, 15)</item>
        ///     <item>expiration [long integer, default null]: time interval in seconds between the due date and the expiration date. ex 123456789</item>
        ///     <item>user [Organization/Project object]: Organization or Project object. Not necessary if StarkBank.Settings.User was set before function call</item>
        /// </list>
        /// <br/>
        /// Return:
        /// <list>
        ///     <item>target Invoice with updated attributes</item>
        /// </list>
        /// </summary>
        public static Invoice Update(string id, string status = null, long? amount = null, DateTime? due = null, long? expiration = null, User user = null)
        {
            (string resourceName, Utils.Api.ResourceMaker resourceMaker) = Resource();
            return Utils.Rest.PatchId(
                resourceName: resourceName,
                resourceMaker: resourceMaker,
                id: id,
                payload: new Dictionary<string, object> {
                    { "amount", amount },
                    { "expiration", expiration },
                    { "status", status },
                    { "due", new Utils.StarkBankDateTime(due) },
                },
                user: user
            ) as Invoice;
        }

        /// <summary>
        /// Retrieve a specific Invoice QRCode png
        /// <br/>
        /// Receive a single Invoice QRCode in png format generated in the Stark Bank API by the invoice ID.
        /// <br/>
        /// <list>
        /// Parameters(required):
        ///     <item>id[string]: object unique id. ex: "5656565656565656"</item>
        /// </list>
        /// <br/>
        /// Parameters(optional) :
        /// <list>
        ///     <item>size [integer, default 7.0]: number of pixels in each "box" of the QR code. Minimum = 1, maximum = 50. ex: 12</item>
        ///     <item>user [Organization/Project object]: Organization or Project object. Not necessary if StarkBank.Settings.User was set before function call</item>
        /// </list>
        /// <br/>
        /// Return:
        /// <list>
        ///     <item>Boleto pdf file</item>
        /// </list>
        /// </summary>
        public static byte[] Qrcode(string id, int? size = null, User user = null)
        {
            (string resourceName, Utils.Api.ResourceMaker resourceMaker) = Resource();

            return Utils.Rest.GetQrcode(
                resourceName: resourceName,
                resourceMaker: resourceMaker,
                id: id,
                options: new Dictionary<string, object> {
                    { "size", size }
                },
                user: user
            );
        }

        internal static (string resourceName, Utils.Api.ResourceMaker resourceMaker) Resource()
        {
            return (resourceName: "Invoice", resourceMaker: ResourceMaker);
        }

        internal static Utils.Resource ResourceMaker(dynamic json)
        {
            long amount = json.amount;
            string name = json.name;
            string taxID = json.taxId;
            string dueString = json.due;
            DateTime? due = Utils.Checks.CheckDateTime(dueString);
            long? expiration = json.expiration;
            double? fine = json.fine;
            double? interest = json.interest;
            List<string> tags = json.tags.ToObject<List<string>>();
            List<Dictionary<string, object>> descriptions = json.descriptions.ToObject<List<Dictionary<string, object>>>();
            List<Dictionary<string, object>> discounts = json.discounts.ToObject<List<Dictionary<string, object>>>();
            foreach (Dictionary<string, object> discount in discounts) {
                discount["due"] = Utils.Checks.CheckDateTime((string)discount["due"]);
            }

            long? nominalAmount = json.nominalAmount;
            long? fineAmount = json.fineAmount;
            long? interestAmount = json.interestAmount;
            long? discountAmount = json.discountAmount;
            string id = json.id;
            string brcode = json.brcode;
            int fee = json.fee;
            string status = json.status;
            string createdString = json.created;
            string updatedString = json.updated;
            DateTime? created = Utils.Checks.CheckDateTime(createdString);
            DateTime? updated = Utils.Checks.CheckDateTime(updatedString);

            return new Invoice(
                amount: amount, name: name, taxID: taxID, due: due, expiration: expiration, fine: fine,
                interest: interest, tags: tags, descriptions: descriptions, discounts: discounts, nominalAmount: nominalAmount,
                fineAmount: fineAmount, interestAmount: interestAmount, discountAmount: discountAmount,
                id: id, brcode: brcode, fee: fee, status: status, created: created, updated: updated
            );
        }
    }
}
